File: src/common/web/api/session/SessionCommands.ts
import { Command } from "../../core/messaging/Command";
import { CommandReply } from "../../core/messaging/CommandReply";
import { CommandComposer } from "../../core/messaging/composers/CommandComposer";
import { CommandReplyComposer } from "../../core/messaging/composers/CommandReplyComposer";
import { MessageBuilder } from "../../core/messaging/composers/MessageBuilder";
import { Message } from "../../core/messaging/Message";

/**
 * Command to get a session value stored on the server. Requires a ``GetSessionValueReply`` reply.
 *
 * @param key - The value key.
 */
@Message.define("command/session/value/get")
export class GetSessionValueCommand extends Command {
    public readonly key: string = "";

    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, key: string, chain: Message | null = null): CommandComposer<GetSessionValueCommand> {
        return messageBuilder.buildCommand(GetSessionValueCommand, { key: key }, chain);
    }
}

/**
 * Reply to ``GetSessionValueCommand``.
 *
 * @param value - The value or *undefined* if no such value was found.*/
@Message.define("command/session/value/get/reply")
export class GetSessionValueReply extends CommandReply {
    public readonly value: any = undefined;

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: GetSessionValueCommand,
        value: any,
        success: boolean = true,
        message: string = ""
    ): CommandReplyComposer<GetSessionValueReply> {
        return messageBuilder.buildCommandReply(GetSessionValueReply, cmd, success, message, { value: value });
    }
}

/**
 * Command to store a session value on the server. Requires a ``SetSessionValueReply`` reply.
 *
 * @param key - The value key.
 * @param value - The value to store.
 */
@Message.define("command/session/value/set")
export class SetSessionValueCommand extends Command {
    public readonly key: string = "";
    public readonly value: any = undefined;

    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, key: string, value: any, chain: Message | null = null): CommandComposer<SetSessionValueCommand> {
        return messageBuilder.buildCommand(SetSessionValueCommand, { key: key, value: value }, chain);
    }
}

/**
 * Reply to ``SetSessionValueCommand``.
 *
 * @param settings - The new user settings (note that these might have been adjusted by the server).
 */
@Message.define("command/session/value/set/reply")
export class SetSessionValueReply extends CommandReply {
    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: SetSessionValueCommand,
        success: boolean = true,
        message: string = ""
    ): CommandReplyComposer<SetSessionValueReply> {
        return messageBuilder.buildCommandReply(SetSessionValueReply, cmd, success, message);
    }
}

File: src/common/py/api/session/session_commands.py
import typing

from ...core.messaging import (
    Command,
    CommandReply,
    Message,
)
from ...core.messaging.composers import (
    MessageBuilder,
    CommandComposer,
    CommandReplyComposer,
)


@Message.define("command/session/value/get")
class GetSessionValueCommand(Command):
    """
    Command to get a session value stored on the server.

    Args:
        key: The value key.

    Notes:
        Requires a ``GetSessionValueReply`` reply.
    """

    key: str

    @staticmethod
    def build(
        message_builder: MessageBuilder, *, key: str, chain: Message | None = None
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(GetSessionValueCommand, chain, key=key)


@Message.define("command/session/value/get/reply")
class GetSessionValueReply(CommandReply):
    """
    Reply to ``GetSessionValueCommand``.

    Args:
        value: The value or *None* if no such value was found.
    """

    value: typing.Any

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: GetSessionValueCommand,
        *,
        value: typing.Any,
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            GetSessionValueReply,
            cmd,
            success,
            message,
            value=value,
        )


@Message.define("command/session/value/set")
class SetSessionValueCommand(Command):
    """
    Command to store a session value on the server.

    Args:
        key: The value key.
        value: The value to store.

    Notes:
        Requires a ``SetSessionValueReply`` reply.
    """

    key: str
    value: typing.Any

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        key: str,
        value: typing.Any,
        chain: Message | None = None,
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(
            SetSessionValueCommand, chain, key=key, value=value
        )


@Message.define("command/session/value/set/reply")
class SetSessionValueReply(CommandReply):
    """
    Reply to ``SetSessionValueCommand``.
    """

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: SetSessionValueCommand,
        *,
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            SetSessionValueReply,
            cmd,
            success,
            message,
        )
